### ISSUE TITLE
[Bug]: Neither Node.js nor NVM is installed when installing Herd

### ISSUE THREAD

**Issue by ericmp33** [beyondcode/herd-community#762]
### Platform

Windows

### Operating system version

W11 23H2 (OS Build 22631.3593)

### System architecture

Intel (x86)

### Herd Version

1.6.1

### PHP Version

PHP 8.3.7 (cli) (built: May  8 2024 08:56:34) (NTS Visual C++ 2019 x64)

### Bug description

I installed Herd. In the docs says [it's installed by default](https://herd.laravel.com/docs/windows/1/advanced-usage/node-versions#managing-nodejs-versions) & it says that [I can checkout what it has installed](https://herd.laravel.com/docs/windows/1/getting-started/installation#installation).

But when checking for `nodejs`:
![image](https://github.com/beyondcode/herd-community/assets/60661635/bf3a1023-6077-495a-8c12-caea202b8316)

Tried to search for errors/logs but idk how to fix this.
Also, the `C:\Users\eric\.config\herd\bin\nvm` folder is empty.

Trying to run `nvm`:
![image](https://github.com/beyondcode/herd-community/assets/60661635/4c01c5e7-467b-4ef1-8ccc-1d9c0fb2e5f0)

Trying to install `nodejs`:
I just click, i wait for some seconds and it doesn't do anything:
![image](https://github.com/beyondcode/herd-community/assets/60661635/e8828d8a-45fd-4b5e-af38-7921549f5e47)

**Comment by ericmp33** [comment]
Update:
Seems I found more logs:
`C:\Users\eric\AppData\Roaming\Herd\logs\main.log`

When I click to install a specific `nodejs` version i get this log:
```
[2024-05-23 19:26:22.297] [info]  [NodeVersionManager] installedNodeVersions: '"C:\\Users\\eric\\.config\\herd\\bin\\nvm\\nvm.exe"' is not recognized as an internal or external command,
operable program or batch file.

[2024-05-23 19:26:22.297] [info]  [NodeVersionManager] parseNodeVersionList: [
  `'"C:\\\\Users\\\\eric\\\\.config\\\\herd\\\\bin\\\\nvm\\\\nvm.exe"' is not recognized as an internal or external command,\r`,
  'operable program or batch file.\r',
  ''
]
[2024-05-23 19:26:22.298] [info]  [NodeVersionManager] installedNodeVersions: 
[2024-05-23 19:26:22.298] [info]  [NodeVersionManager] parseNodeVersionList: [ '' ]
[2024-05-23 19:26:23.300] [info]  [NodeVersionManager] install: 'C:\Users\eric\.config\herd\bin\nvm\nvm.exe' is not recognized as an internal or external command,
operable program or batch file.

[2024-05-23 19:26:23.300] [info]  [NodeVersionManager] install: 
[2024-05-23 19:26:24.116] [info]  [NodeVersionManager] installedNodeVersions: '"C:\\Users\\eric\\.config\\herd\\bin\\nvm\\nvm.exe"' is not recognized as an internal or external command,
operable program or batch file.

[2024-05-23 19:26:24.116] [info]  [NodeVersionManager] parseNodeVersionList: [
  `'"C:\\\\Users\\\\eric\\\\.config\\\\herd\\\\bin\\\\nvm\\\\nvm.exe"' is not recognized as an internal or external command,\r`,
  'operable program or batch file.\r',
  ''
]
[2024-05-23 19:26:24.117] [info]  [NodeVersionManager] installedNodeVersions: 
[2024-05-23 19:26:24.117] [info]  [NodeVersionManager] parseNodeVersionList: [ '' ]
```

**Comment by sschlein** [comment]
That's odd. Can you check your environment variables if there are some from a previous NVM installation? They might also be in your path.

If you remove them and restart Herd, you should see an option to install nvm again.

**Comment by ericmp33** [comment]
Yeah, well I'll explain the trip i faced.

So yeah I already checked the env variables & everywhere where could be residual `nodejs`/`npm`/`nvm` stuff (Some time ago I installed `nodejs` on the computer but I uninstalled along with PHP since at that point I was migrating to Herd so there was no point on having those manual installations - Herd would manage that for me).

Also restarted my computer & reinstalled Herd several times in different ways.

After a lot of unsuccessful tries I said, you know what, I'm gonna install `nvm` manually. So I downloaded the exe `v1.1.12`. Then I installed it manually into `C:\Users\eric\.config\herd\bin\nvm`.

Next, I opened Herd & tried to install the latest node version, but I got another issue (https://github.com/coreybutler/nvm-windows/issues/1068):
```
NVM for Windows should be run from a terminal such as CMD or PowerShell
```

So I again uninstalled all `nodejs`/`npm`/`nvm` stuff, paths, everything. Just in case. And I installed the `v1.1.11`.

After that, I opened Herd & tried to install the latest `node` version. It worked. No logs, no errors. So now my current `node` version is the latest (the checkbox where it says it's the current node version was checked).

Despite of that, when I tried to do `npm i` in the project I was developing I got:
```
npm is not recognized as a name of a cmdlet, function, script file, or executable program.
```

I said, WTF. But anyways I knew that I cannot actually trust the Herd's GUI in the Windows version. Somehow, in my experience using it, I've already had issues related to GUI saying A when CLI was saying B.

I went to [https://herd.laravel.com/docs/windows/1/advanced-usage/node-versions#via-the-cli](https://herd.laravel.com/docs/windows/1/advanced-usage/node-versions#via-the-cli) and I did `nvm use v21.7.3`.

And that was the last step, I finally did `npm i` and just worked. 🎉

So I have no clue what is happening internally but I solved it this way, installing `nvm` manually, then installing `node` via Herd's GUI & finally, via CLI, doing `nvm use v21.7.3`.

I hope this can help someone facing this same issue!
And if someone has a better workaround/fix for this, please share it!

**Comment by khanra17** [comment]
removing `NVM_HOME` & `NVM_SYMLINK` -> then installing Herd, solved the issue for me.
---
![image](https://github.com/beyondcode/herd-community/assets/65936206/48f00676-f6cc-4a74-82fe-ebefe58db1e3)

**Comment by ericmp33** [comment]
@sschlein cool it worked for you. I also did it but in my case didn't solve it 🤷

**Comment by halmanza** [comment]
Did you remove only from user variables ? @ericmp33 . I also had to remove these from the system variables section as well.

**Comment by ericmp33** [comment]
@halmanza I removed every reference to node, nvm & npm variables from my user & from the system too

**Comment by khanra17** [comment]
I installed and uninstalled Herd several times without issue.

But once, I downloaded Node directly (Herd was not installed at that time) and uninstalled Node. Then, when trying to install Herd again, Node was not being installed from Herd.

I uninstalled Herd, cleaned `ProgramData/`, `AppData/`, and every possible folder, and cleaned `User & System PATHS`. Nothing worked.

Then, I noticed `NVM_HOME` and `NVM_SYMLINK`. After removing them, I reinstalled Herd and saw the PowerShell popup. After finishing, it indeed installed Node, and everything works!

**Comment by sschlein** [comment]
@khanra17 So our of curiousity: Why do you install and uninstall Herd several times in the first place?

**Comment by sschlein** [comment]
This seems to be resolved and so I am closing this issue.

TL;DR;
Herd checks for NVM environment variables in your system and user path. If they are present, it does not install it and assumes that it is already available. For the initial issue, there seems to be a different problem because Herd created the "nvm" directory but did not add any files there, so there might be an issue with the internet connection during the setup – after that, it had an nvm directory and variables in the paths but no actual binaries and that caused the issue.

**Comment by ericmp33** [comment]
If you talk about my case, I didn't have internet connection issues during the installation of it 🤷

**Comment by sschlein** [comment]
Yup, but maybe nvm wasn't available for that time.

**Comment by ericmp33** [comment]
Okay, if I have time I'll uninstall Herd once more and install it again to see what happens. Then I'll let you know how it went here ✌️

**Comment by sschlein** [comment]
If you do a clean install, can you use this version with more logging?

[Download](https://herd-dev.fra1.digitaloceanspaces.com/test-builds/Herd-1.6.1-latest.exe.zip)

Get the logs from %USERPROFILE%\AppData\Roaming\Herd\logs

You can either post the logs here or send them to [support@beyondco.de](mailto:support@beyondco.de)

Thanks for helping us with debugging.

**Comment by khanra17** [comment]
> @khanra17 So our of curiousity: Why do you install and uninstall Herd several times in the first place?

Installed when needed for laravel project, uninstalled when project finished.
Earlier was using docker for php, mysql so some backnforth

**Comment by DannJohnrem** [comment]
I have figure out a solution:

if you have a nvm not connected to herd just uninstall it and restart the laravel herd, after doing that go to 'node' there is a nvm installer button there.

![image](https://github.com/user-attachments/assets/76244f83-ddbe-4c9b-8bdf-a592b4f544f0)

**Comment by denifelixe** [comment]
Follow the complete uninstall instruction here: https://herd.laravel.com/docs/windows/troubleshooting/uninstalling

then reinstall Herd.

It works for me!

**Comment by anttiryt** [comment]
I had a bit similar. Herd opened terminal didn't recognize nvm or node. The needed lines were in .zshrc but seems .zshrc is not ran. 

Running **zsh -l** in the terminal brought the needed Herd extensions

**Comment by Al-Pharaday** [comment]
Ran into a similar issue with Herd on Windows 11 tday. The `...\.config\herd\bin\nvm` had nothing but `mvm.exe` and a couple `elevate` files after installation. It is a 100% Herd's issue, as they say zsh is sort of opinionated. And I guess they do not use nvm4windows.

I opted for a straightforward approach:
1. installed nvm manually (https://github.com/coreybutler/nvm-windows  - it places the nvm files in `C:\Users\<user>\AppData\Local\nvm` by default, it is NVM_HOME now)
2. Deleted the `...\.config\herd\bin\nvm` folder
3. Created a junction `mklink /D C:\Users\<user>\.config\herd\bin\nvm C:\Users\<user>\AppData\Local\nvm`
4. Restarted Herd

**Comment by nannnmuzaki** [comment]
### i think i found the solution!

After you deleted the `NVM_HOME` and `NVM_SYMLINK` environment variables, make sure that you **open the Herd app as an administrator**.

**Comment by oneil-eliervin** [comment]
I had the same issue as @Al-Pharaday on a fresh Windows 11 installation. Everything seemed to be working except nvm wasn't getting installed (and thus neither npm/node). 

In my case, this is likely due to the enterprise environment I'm in requiring Powershell scripts to be signed so the install command was having issues:

`Start-Process PowerShell -ArgumentList '-NoProfile -ExecutionPolicy Bypass -File "C:\Program Files\Herd\resources\app.asar.unpacked\resources\bin\nvm\install.ps1" -NVM_PATH "C:\Users\<username>\.config\herd\bin\nvm"' -Verb RunAs -Wait`

Following the approach @Al-Pharaday did with nvm-windows got me up and running and let the Herd UI handle the Node installs

One additional note is the main.log file contained a few nvm settings related errors as well:

`[2025-09-18 16:05:58.601] [info]  [NodeVersionManager] parseNodeVersionList: [
  '',
  'ERROR open C:\\Users\\<username>\\.config\\herd\\bin\\nvm\\settings.txt: The system cannot find the file specified.',
  ''
]`

**Comment by shaffe-fr** [comment]
I resolved a similar issue by doing the following:

1. Opened `~/.config/bin/nvm` in PowerShell
2. Copied `settings.txt` to `~\AppData\Roaming\nvm`
3. Ran `nvm use x.x.x`

**Comment by GregOstry** [comment]
Workaround for Windows 11 25H2 Enterprise (Hardened Environment)

I ran into the same issue on Windows 11 25H2 Enterprise.
This version appears to be significantly more hardened, and PowerShell script execution is restricted at a deeper security level (even when ExecutionPolicy reports Bypass).

As a result, Herd’s internal install.ps1 for NVM cannot be executed because unsigned scripts in Program Files are blocked by enterprise security controls (Code Integrity / CI policies).

Thanks to @Al-Pharaday for pointing in the right direction — here is a clean and stable workaround that avoids modifying any enterprise security policies.

1) Close Herd completely
2) Install NVM for Windows
       Download and install: [https://github.com/coreybutler/nvm-windows ](https://github.com/coreybutler/nvm-windows)
       Use the default installation path (typically): `C:\Users\<your-user>\AppData\Local\nvm`

3) Remove Herd’s bundled NVM directory
      Open PowerShell as Administrator and run: `Remove-Item -Recurse -Force "$env:USERPROFILE\.config\herd\bin\nvm"`

4) Create a directory junction
     `cmd /c mklink /D "$env:USERPROFILE\.config\herd\bin\nvm" "$env:LOCALAPPDATA\nvm"`

5) Open Herd again
     Now you can install Node versions normally via the Herd UI.
